outlines/blaze.md

```diff
@@ -0,0 +1,56 @@
+# Blaze
+
+1. Introduction to Blaze -- the tracker-backed handlebars like templating syntax
+  1. Example of spacebars syntax, data context + helpers
+  2. Data contexts and access - name lookup (`.`, `this`, `..`, `[0]`), and null values
+  3. Helpers, arguments, options
+  4. Inclusion of other templates + arguments, passing data contexts
+  5. Helpers inside tags -- returning strings and objects, `checked={{isChecked}}`
+  6. Nested helpers + sub expressions
+  7. Block helpers (you can create them with templates, see 10. below)
+  8. Safestrings and `{{{`
+  9. Builtin block helpers
+    1. `{{#if/unless}}`
+    2. `{{#with}}`
+    3. `{{#each}}`
+    3. `{{#let}}`
+  10. Comments
+  11. Strictness
+  12. Escaping
+2. Creating reusable "pure" components with Blaze / best practice (a lot of this is repeating @sanjo's boilerplate)
+  1. Validating data context fits a schema
+  2. Always set data contexts to `{name: doc}` rather than just `doc`. Always set a d.c on an inclusion.
+  3. Use `{{#each .. in .. }}` to achieve the above
+  4. Use the template instance as a component -- adding a `{{template}}` helper to access it
+  5. Use a (named / scoped on _id if possible) reactive dict for instance state
+  6. Attach functions to the template instance (in `onCreated`) to sensibly modify state
+  7. Place `const template = Template.instance()` at the top of all helpers that care about state
+  8. Always scope DOM lookups with `this.$`
+  9. Use `.js-X` in event maps
+  10. Pass extra content to components with `Template.contentBlock`, or named template arguments
+  11. Use the `onRendered` callback to integrate w/ 3rd party libraries
+3. Writing "smart" components with Blaze
+  1. All of section 2.
+  2. Use `this.autorun` and `this.subscribe`, listening to `FlowRouter` and `this.state`
+  3. Set up cursors in helpers to be passed into pure sub-components, and filtered with `cursor-utils`
+  4. Access stores directly from helpers.
+4. Reusing code between templates
+  1. Prefer utilities/composition to mixins or inheritance
+  2. How to write global helpers
+5. Understanding Blaze
+  1. When does a template re-render (when its data context changes)
+  2. When does a helper-rerun? (when its data context or reactive deps change)
+    1. So be careful, this can happen a lot! If your helper is expensive, consider something like https://github.com/peerlibrary/meteor-computed-field
+  3. How does an each tag re-run / decide when new data should appear?
```

Comment:
So DOM element attributes. Things like `<div {{attrs}}>`. So we should explain how do they reactivelly change as well, no?

BTW, with Blaze Components there is also a nice pattern which emerged for attributes. Namely that you can have multiple sources contribute to the same dict which you then render out. Like a dict of CSS classes, and inline style, and then you render them at once out. Instead of patching them together in templates. With inheritance and mixins you can do this really well.